File: AnyTorah/API/EinAyahLoader.swift
import Foundation

/// Loads the bundled Ein Ayah JSON and returns entries for a specific Talmud daf.
///
/// JSON structure: {"berakhot": {"2a": [{"citation": "...", "text": "..."}], ...}, "shabbat": {...}}
/// Ein Ayah only comments on aggadic passages, so many dafim have no entries.
struct EinAyahLoader {

    private static let cache: [String: Any] = {
        guard let url = Bundle.main.url(forResource: "ein_ayah", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return [:] }
        return json
    }()

    /// `daf` is a bare daf number string (e.g. "3") — the app navigates whole dafim.
    /// Combines entries from amud a and amud b.
    static func entries(tractate: String, daf: String) -> [CommentaryEntry] {
        let key = tractateKey(for: tractate)
        guard let dafMap = cache[key] as? [String: Any] else { return [] }

        let rawA = (dafMap["\(daf)a"] as? [[String: String]]) ?? []
        let rawB = (dafMap["\(daf)b"] as? [[String: String]]) ?? []
        let combined = rawA + rawB

        return combined.enumerated().compactMap { idx, raw in
            guard let text = raw["text"], !text.isEmpty else { return nil }
            let citation = raw["citation"] ?? ""
            let he = citation.isEmpty ? text : "\(citation)\n\(text)"
            return .text(index: idx, he: he, en: "")
        }
    }

    private static func tractateKey(for sefariaName: String) -> String {
        switch sefariaName.lowercased() {
        case "berakhot":  return "berakhot"
        case "shabbat":   return "shabbat"
        default:          return sefariaName.lowercased()
        }
    }
}

File: AnyTorahAndroid/app/src/main/java/com/anytorah/api/EinAyahLoader.kt
package com.anytorah.api

import android.content.Context
import com.anytorah.models.CommentaryEntry
import org.json.JSONObject

/**
 * Loads the bundled Ein Ayah JSON and returns entries for a specific Talmud daf.
 *
 * JSON structure: {"berakhot": {"2a": [{"citation": "...", "text": "..."}], ...}, "shabbat": {...}}
 * Ein Ayah only comments on aggadic passages, so many dafim have no entries.
 */
object EinAyahLoader {

    private var cache: JSONObject? = null

    private fun getCache(context: Context): JSONObject {
        cache?.let { return it }
        val json = context.assets.open("ein_ayah.json").bufferedReader().use { it.readText() }
        return JSONObject(json).also { cache = it }
    }

    /// `daf` is a bare daf number string (e.g. "3") — the app navigates whole dafim.
    /// Combines entries from amud a and amud b.
    fun entries(context: Context, tractate: String, daf: String): List<CommentaryEntry> {
        val data = try { getCache(context) } catch (e: Exception) { return emptyList() }
        val key = tractateKey(tractate)
        val dafMap = data.optJSONObject(key) ?: return emptyList()

        val result = mutableListOf<CommentaryEntry>()
        var idx = 0
        for (amud in listOf("${daf}a", "${daf}b")) {
            val arr = dafMap.optJSONArray(amud) ?: continue
            for (i in 0 until arr.length()) {
                val entry = arr.optJSONObject(i) ?: continue
                val text = entry.optString("text").takeIf { it.isNotEmpty() } ?: continue
                val citation = entry.optString("citation", "")
                val he = if (citation.isEmpty()) text else "$citation\n$text"
                result.add(CommentaryEntry.Text(index = idx++, he = he, en = ""))
            }
        }
        return result
    }

    private fun tractateKey(sefariaName: String): String = when (sefariaName.lowercase()) {
        "berakhot" -> "berakhot"
        "shabbat"  -> "shabbat"
        else       -> sefariaName.lowercase()
    }
}
